src/librustc_mir/transform/instrument_coverage.rs

```diff
@@ -24,46 +27,96 @@ pub struct InstrumentCoverage;
-                                    num_counters = std::cmp::max(num_counters, index + 1);
-                                }
-                            }
-                        }
-                    }
+    providers.coverage_data = |tcx, def_id| coverage_data_from_mir(tcx, def_id);
+}
+
+fn coverage_data_from_mir<'tcx>(tcx: TyCtxt<'tcx>, mir_def_id: DefId) -> &'tcx CoverageData {
+    let mir_body = tcx.optimized_mir(mir_def_id);
+    // FIXME(richkadel): The current implementation assumes the MIR for the given DefId
+    // represents a single function. Validate and/or correct if inlining and/or monomorphization
+    // invalidates these assumptions.
+    let count_code_region_fn = tcx.require_lang_item(lang_items::CountCodeRegionFnLangItem, None);
+
+    // The `num_counters` argument to `llvm.instrprof.increment` is the number of injected
+    // counters, with each counter having an index from `0..num_counters-1`. MIR optimization
+    // may split and duplicate some BasicBlock sequences. Simply counting the calls may not
+    // not work; but computing the num_counters by adding `1` to the highest index (for a given
+    // instrumented function) is valid.
+    let mut indexed_regions = vec![];
+    let mut num_counters: u32 = 0;
+    //for terminator in count_code_region_terminators(tcx, mir_body) {
+    for terminator in traversal::preorder(mir_body)
+        .map(|(_, data)| (data, count_code_region_fn))
+        .filter_map(count_code_region_terminator_filter)
+    {
+        // FIXME(richkadel): When analyzing the MIR and identifying CodeRegions to count by LLVM
+        // counter_expressions, inject (and here, filter for) an additional call, such as:
+        //
+        //     code_region_counter_expression(
+        //         left_counter: i32, signed_right_counter: i64,
+        //         start_byte_pos, end_byte_pos
+        //     )
+        //
+        // `abs(signed_right_counter)` (perhaps there's a better name?) is the index of the counter's
+        // count to be added to or subtracted from (depending on the sign) the `left_counter`
+        // index.
+        //
+        // Then store the expressions in a different IndexVec.
+        // We may also want to consider "GapRegions" (clang had them... do we need them too?)
+        if let TerminatorKind::Call { args, .. } = &terminator.kind {
+            let (index, start_byte_pos, end_byte_pos) = count_code_region_args(&args);
+            indexed_regions.push((index, CoverageRegion { start_byte_pos, end_byte_pos }));
+            num_counters = std::cmp::max(num_counters, index + 1);
+        }
+    }
+    let hash = if num_counters > 0 { hash_mir_source(tcx, mir_def_id) } else { 0 };
+    let mut coverage_regions = IndexVec::<u32, CoverageRegion>::from_elem_n(
+        CoverageRegion { start_byte_pos: 0, end_byte_pos: 0 },
+        indexed_regions.len(),
+    );
+    for (index, region) in indexed_regions {
+        coverage_regions[index] = region;
+    }
```

Comment:
Cool. Thanks for the suggestion. I wasn't 100% sure `from_fn_n()` was smart enough to allocate the underlying array only once, but after investigating a bit more, it looks like it is smart enough.

I think it would be helpful to make a note of that in the from_fn_n() doc comment for IndexVec(). WDYT? If so, is it worth creating a whole new PR for that, or just slipping it into this one?

Anyway, as it turns out, with the updates I just made, this code has been removed.

Reply:
FYI, I updated went ahead and updated the `from_fn_n` doc comment in the latest commit.

Reply:
Thank you!